internal/controllers/gateway/gateway_utils.go

```diff
@@ -265,13 +265,65 @@ func getListenerStatus(
 		if listener.Hostname != nil {
 			hostname = *listener.Hostname
 		}
+		supportedkinds, ResolvedRefsReason := getListenerSupportedRouteKinds(listener)
+
+		// If the listener uses TLS, we need to ensure that the gateway is granted to reference
+		// all the secrets it references
+		if listener.TLS != nil {
+			TLSresolvedRefReason := string(gatewayv1beta1.ListenerReasonResolvedRefs)
+			for _, certRef := range listener.TLS.CertificateRefs {
+				// if the certificate is in the same namespace of the gateway, no ReferenceGrant is needed
+				if certRef.Namespace != nil && *certRef.Namespace != (Namespace)(gateway.Namespace) {
+					// get the result of the certificate reference. If the returned reason is not successful, the loop
+					// must be broken because the secret reference isn't granted
+					TLSresolvedRefReason = getReferenceGrantConditionReason(gateway.Namespace, certRef, referenceGrants)
+					if TLSresolvedRefReason != string(gatewayv1beta1.ListenerReasonResolvedRefs) {
+						break
+					}
+				}
+
+				// only secrets are supported as certificate references
+				if (certRef.Group != nil && (*certRef.Group != "core" && *certRef.Group != "")) ||
+					(certRef.Kind != nil && *certRef.Kind != "Secret") {
+					TLSresolvedRefReason = string(gatewayv1beta1.ListenerReasonInvalidCertificateRef)
+					break
+				}
+				secret := &corev1.Secret{}
+				secretNamespace := gateway.Namespace
+				if certRef.Namespace != nil {
+					secretNamespace = string(*certRef.Namespace)
+				}
+				if err := client.Get(ctx, types.NamespacedName{Namespace: secretNamespace, Name: string(certRef.Name)}, secret); err != nil {
+					if !apierrors.IsNotFound(err) {
+						return nil, err
+					}
+					TLSresolvedRefReason = string(gatewayv1beta1.ListenerReasonInvalidCertificateRef)
+				}
+			}
+			if gatewayv1beta1.ListenerConditionReason(TLSresolvedRefReason) != gatewayv1beta1.ListenerReasonResolvedRefs {
+				ResolvedRefsReason = gatewayv1beta1.ListenerConditionReason(TLSresolvedRefReason)
+			}
+		}
+
 		status := ListenerStatus{
 			Name:           listener.Name,
 			Conditions:     []metav1.Condition{},
-			SupportedKinds: getListenerSupportedRouteKinds(listener),
+			SupportedKinds: supportedkinds,
 			// this has been populated by initializeListenerMaps()
 			AttachedRoutes: listenerToAttached[listener.Name],
 		}
+
+		// if the resolvedRefs condition is not successful, append the resolvedRefs condition failed with the proper reason
+		if ResolvedRefsReason != gatewayv1beta1.ListenerReasonResolvedRefs {
+			status.Conditions = append(status.Conditions, metav1.Condition{
```

Comment:
Why `append`? Could we guarantee that their must not be existing `ListenerConditionResolvedRefs` conditions?

Reply:
yes, we are building the status from sratch here